### PR TITLE
Add strict opts on decode

### PR DESCRIPTION
**Pull request by soteras**
#86 

I've refactor **Decode** module, because when I tried to add new opt is was very difficult because all functions add key_decode, string_decode and I would have to refactor all functions, now It's ease to add new options. 

I added strict option on decode because I need to know when my JSON has duplicated keys. 

```elixir
iex(3)> Jason.decode!(~s({"a": 1, "a": 2}), maps: :strict)
** (Jason.DecodeError) duplicate key: a
    (jason) lib/jason.ex:84: Jason.decode!/2
```

### CODE REVIEW

File: lib/decoder.ex [hunk 1]
```diff
@@ -41,285 +47,318 @@ defmodule Jason.Decoder do
...
+  defp key_decode_function(%{keys: :atoms}), do: &String.to_atom/1
+  defp key_decode_function(%{keys: :atoms!}), do: &String.to_existing_atom/1
+  defp key_decode_function(%{keys: :strings}), do: &(&1)
+  defp key_decode_function(%{keys: fun}) when is_function(fun, 1), do: fun
+
+  defp decode_map_function(%{maps: maps}) do
+    case maps do
+      :strict -> &map_strict/2
+    end
+  end
```
Comment (paulobezerr) [inline lib/decoder.ex]: I think that this can throw "no case clause matching"

**Comment by michalmuskala** [comment]
Unfortunately, I'm afraid that extracting the options and allocating the funs on each element is going to kill the performance completely. In this library in particular performance is far more important than ease of maintenance. That said, I haven't actually benchmarked it, it might be fine - loosing 1 to 2% would be ok, anything more is probably prohibitive.

**Comment by soteras** [comment]
Is there a way to benchmarked it?

**Comment by michalmuskala** [comment]
There are benchmarks in the `bench` folder with some instructions in the README.

I run some trimmed down version of the full benchmark and right now it seems that on object-heavy JSON this PR is about 4-7% slower than current master without strict decoding.

**Comment by michalmuskala** [comment]
Given no changes in status of this PR for over a year, I'm going to close it. If there's a desire to pick up the work again, please re-open this with the required changes & benchmark data.

**Pull request closed without merging**